src/librbd/cache/pwl/Types.h

```diff
@@ -299,6 +299,7 @@ struct WriteLogPoolRoot {
   uint32_t num_log_entries;
   uint32_t first_free_entry;     /* Entry following the newest valid entry */
   uint32_t first_valid_entry;    /* Index of the oldest valid entry in the log */
+  char id[36];                   /* id for this open request */
```

Comment:
And shouldn't it be `id[37]` to include the nul char? Because uuid print function you use to fill the buffer uses `memcpy(..., 37)` [1].

[1] https://github.com/ceph/ceph/blob/7c7eb251305f6e337d19df8319dee84dfdd89c99/src/include/uuid.h#L50

Reply:
I just don't know how to use `string` here. The implementation of `string` is that the pointer points to a memory area, but this data structure needs to be persisted to pmem or SSD. Isn't there an error when reopening and loading `string`?

The 37th char of uuid.print() is '\0'. It needs to be discarded, otherwise it will be inconsistent when reconstructing the string for comparison

Reply:
> I just don't know how to use string here

As I understood from some our previous conversation the plan is to switch to ceph standard encode/decode for on-disk format. In that case it will be easy to use string. Until then I suppose char buf is good enough.

> The 37th char of uuid.print() is '\0'. It needs to be discarded, otherwise it will be inconsistent when reconstructing the string for comparison

If you think the `uuid.print()` does wrong then it should be fixed (as a separate commit). If `uuid.print()` behavior remains the same then I think you need to increase your buffer (or does not use `uuid.print()`, just call `memcpy` itself) to avoid invalid memory access. 

To me having the buffer one extra byte to store the nul character would not hurt and actually would be safer.

Reply:
I see now. I made a mistake. I should set it to 37. Otherwise, `‘\0’` will be written to the unknown area, but only 36 characters need to assign() to string when constructing string comparison

Reply:
Sorry, I didn't quite understand what you meant about using string first. Jianpeng explained it to me, If use encode/decode, `string` can be used here. So I add a TODO, change it better in the future. RWL should also use encode/decode mode.